plugins/workspace-plugin/src/workspace-folder-updater.ts

```diff
@@ -0,0 +1,91 @@
+/**********************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as theia from '@theia/plugin';
+const UPDATE_WORKSPACE_FOLDER_TIMEOUT = 5000;
+
+export class WorkspaceFolderUpdater {
+  private pendingFolders: string[] = [];
+  private addingWorkspaceFolderPromise: Promise<void> | undefined;
+
+  async addWorkspaceFolder(path: string): Promise<void> {
+    const workspaceFolderPath = this.toValidWorkspaceFolderPath(path);
+    if (this.pendingFolders.includes(workspaceFolderPath)) {
+      return Promise.resolve();
+    }
+
+    if (this.addingWorkspaceFolderPromise) {
+      this.pendingFolders.push(workspaceFolderPath);
+    } else {
+      try {
+        this.addingWorkspaceFolderPromise = this.addFolder(workspaceFolderPath);
+        await this.addingWorkspaceFolderPromise;
+      } catch (error) {
+        console.error(error);
```

Comment:
Why do we need to handle this here and not let the error propagate?

Reply:
If a request for adding a folder failed - we should continue to add other folders.